### PR TITLE
fix: free registration memory between tiles, skip the worst-case scan without a free axis

### PR DESCRIPTION
**Pull request by vboussot**
## Description

Two fixes left over from the tiled registration work, uncommitted since #184 and #189 merged.

- **Each registration's memory is freed before the next tile** (`impact-reg`).
  - FireANTs keeps its CUDA tensors in reference cycles that only Python's cyclic collector frees. That collector runs on allocation counts, not device memory, so several tiles' worth of fields stayed allocated and a tiled run ran out of the card. `FireANTsRegistration` now collects after each call.
  - ConvexAdam's progress observer holds the fine filter, and the filter holds the observer: a cycle through C++ that the collector cannot see, so every call kept its filter and its fields alive. The observers are removed once `Update()` returns, errors included.
- **No worst-case scan without a free patch axis** (`vram`). `_presize_free_axes` asked the dataset for its worst case before checking there was anything to size. That scan takes every group's extents, non-image groups included (fiducials, transforms), whose shapes have no spatial axes to compare. It now returns early when no axis is free.

The empty-fixed-mask change that sat in the same working tree is not here: #189 already merged it.

## Type of change

- [x] `fix`: bug fix

## How has this been tested?

```bash
ruff check && ruff format --check        # clean on the touched files
python -m pytest tests/unit/test_vram_sizing.py apps/impact_reg/tests/unit -q   # 91 passed, dev environment
```

The vram guard has a unit test, which fails on `main`. The two memory fixes have no test: both need a GPU run over many tiles, and this PR does not include a before/after measurement of device memory.

## Checklist

- [x] PR title and commits follow **Conventional Commits**
- [ ] `pixi run check`: not run in full, only ruff and the tests above
- [ ] konfai-apps tests: impact_reg unit tests only

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved registration reliability by ensuring temporary resources are released after processing, including when errors occur.
  - Reduced unnecessary memory usage during repeated registration operations.
  - Improved VRAM planning for fixed-size or non-image data by skipping unnecessary shape analysis when no adjustable axes are available.

- **Tests**
  - Added regression coverage for VRAM planning when no free axes exist.

### CODE REVIEW

File: apps/impact_reg/impact_reg_konfai/models/fireants.py [hunk 1]
```diff
@@ -812,6 +813,10 @@ def forward(
                 fixed_mask_img = data_to_image(fixed_mask[b].detach().cpu().numpy(), fmask_attrs[b])
                 moving_mask_img = data_to_image(moving_mask[b].detach().cpu().numpy(), mmask_attrs[b])
                 dvf_np = self._engine.register(fixed_img, moving_img, device_index, fixed_mask_img, moving_mask_img)
+                # FireANTs' registration objects keep their CUDA tensors in reference cycles that only the cyclic
+                # collector frees; it runs on Python allocation counts, not device memory, so without a collection
+                # per call several tiles' worth of fields stay allocated and a tiled run runs out of the card.
+                gc.collect()
```
[the remark this answers is not in the document]
Comment (vboussot) [reply apps/impact_reg/impact_reg_konfai/models/fireants.py]: Fixed in e99c165: the collection now runs in a `finally`, so a failed call (an OOM restart retrying on the same card) frees its cycles too.
